### ISSUE TITLE
Bring all packaging layers into a single fixpoint

### ISSUE THREAD

**Issue by roberth** [NixOS/nixpkgs#273815]
### Describe the problem

This issue proposes a solution to a number of problems.

- Interface for package overriding is inconsistent
- RFC 92 dynamic derivations-based packages will not be representable by `mkDerivation`
- Packages leak their implementation details, without any sign that they are implementation details
- It is unclear whether an attribute is meant to be used by the builder (script) of the derivation
- Isolating the package derivation from documentation tools is not feasible despite multi-output derivations
- It is almost impossible to write composable, reusable pieces of logic that affect multiple arguments of `mkDerivation`
  - `mkDerivation` has significant pressure to grow, become huge, incomprehensible, slow, and mass-rebuild inducing

When packaging, we have to keep six+ layers of attribute sets in mind for various purposes.
That is quite a lot, and it leads to issues when one needs to access information from a particular layer in another layer.
For instance, overriding functions are generally only available for ~3 of the layers, and using some may revert the effect of previous overrides.

As an example, the layers of a Python package are:

- package function args (ie `callPackage`, `.override`)
- mkPythonModule args
- mkDerivation args
- derivation args (technically inaccessible, but may leak to next layer)
- package attributes
- package attributes for a different output (`pkg.dev != pkg`, but is very similar)
- cross splicing

### Proposed solution

The layers themselves are valid, mostly, but the way they are composed, by _"ad hoc" functions_, is the cause of aforementioned problems. Instead, we may compose them in a manner that may be somewhat familiar from the **module system**.

Use of the module system has been explored extensively and successfully by @DavHau and [dream2nix](https://github.com/nix-community/dream2nix). However, Nixpkgs has such a scale that we need to care far more about even a constant factor overhead, such as would be imposed by the module system. Experimentally, we have seen that this pretty much **rules out** wide use of such a feature rich system.

However, that does not mean that we need to reject what I would consider the core features of the module system: a fixpoint of a monoid. What does that mean?

Fixpoint: we declare things using functions, where the argument is the "final" result. This recursion allows access to "variables" or "option" in a way that might feel similar to fields in an object oriented language.
Monoid: we don't use a single function, but multiple, and their results are merged in some way.

In the module system, this merging operation is very elaborate, and therefore a source of evaluation overhead. This must be avoided.
A possible, lighter weight version of this is the [merge operation in minimod](https://gist.github.com/roberth/940dff88ca5f5f95949dc309dbe60a65#file-minimod-nix-L126). An even lighter alternative is that of overlays: little more than `//`, also known as `attrsets.merge`, which is non-recursive and minimally helpful.

And that brings us to the final crucial element, which is **overriding**. The module system uses numeric priority markers to specify which definitions win, whereas with overlays, the last overlay composition operation wins. That latter is more efficient, but harder to use, as any merging needs to be specified by hand, e.g. `buildInputs = o.buildInputs or [] ++ ...` etc.
Finally, we may consider leaving all merging behavior up to the user, and let them pick between such methods. This would be most flexible, but imposes more complexity on package authors.

What might this look like? Assuming we go with a limited amount of merging, and a last-wins overriding system as described, a package might as follows. Some required attributes, such as `setup.name` are omitted.

```nix
mkDerivationPackage({ pythonAttrs, setup, drvAttrs, public, ... }:
  # Don't need to get stuff from four levels deep
  # setup: the generic shell script from "stdenv"
  setup.configureFlags = ... (optional drvAttrs.doCheck "--enable-tests") ...;
  setup.doCheck = true; # like mkDerivation { doCheck } argument
  # No more passthru. It's symmetric now.
  public.tests = callPackage ./tests { mypkg = public; };
)
```

An override may look like:

```nix
pkg.override (self: { setup, ...}: {
  setup.buildInputs = super.setup.buildInputs ++ [ somePkg ];
  derivation.disallowedReferences = [ ];
  # Add a marker that isn't propagated to the derivation.
  public.hasSomePkg = true;
})
```

The "replacement" of `callPackage` may look like

```nix
# top-level.nix (or RFC 140 impl)
pkgs.mkDerivationPackage ../foo.nix;
```

```nix
# ../foo.nix
{ deps, ... }:

{
  deps = { pkgs, ... }: {
    # Some defaults may be obvious
    hello = pkgs.hello;
    # But here we introduce `boost` as a stable identifier to allow overriding without knowledge of the current default attribute.
    boost = pkgs.boost_180;
  };
  setup.buildInputs = [ deps.boost ];
  setup.nativeBuildInputs = [ deps.hello ];
}
```

### Implementation

In order to implement this functionality in a sustainable manner, we need to **disentangle** the `setup -> derivation` transformation that is currently implemented in `make-derivation.nix`, mixed with parts of the implementation of `overrideAttrs`.
Taking this apart will greatly benefit readability, but more importantly allow us to improve the overriding mechanisms without unsustainable code duplication with `mkDerivation`.

The first layer to implement is the package attrs layer. This layer is fairly simple. It is responsible for taking the `public` attribute from the fixpoint and returning it.

The next layer to implement is the `derivation` layer. It writes the result of `builtins.strictDerivation` to `public`, perhaps with minor tweaks. This is more efficient than `builtins.derivation` and it implements a separation between a package's public, supported interface, and attributes that are implementation details. See #217243.

Then the disentangled `mkDerivation` logic can be applied, taking arguments from `setup` (or some other name; TBD), and returning into the `derivation` attribute of the fixpoint. This should not include redundant (and slow)

For the multi-outputs wrapper functionality that presents alternate outputs as full blown package attrsets, I would suggest to provide an alternative attribute `outputs = { bin = "<store path string>"; dev = "<store path string>"; }`, which is sufficient for almost all usages, and more efficient to evaluate. Legacy output attributes may still be provided, but should be avoided by library code and builder functions if `outputs` is an attrset.

I am confident that we can combine all layers except perhaps cross splicing into a single fixed point. Cross splicing does not represent a 1:1 function invocation, so may need to remain similar to today, although I wouldn't exclude the possibility of making improvements in that area.

For the details of the `deps` pattern, refer to dream2nix.

At this point, the `mkDerivation` functionality has been replicated in a better way that solves most of the problems. Porting the language infrastructures to the new style will take further effort. As similar approach to that with `mkDerivation` can be applied.

### Additional context

- A little bit of discussion happened here and in transitively and reversely linked issues
  - https://github.com/NixOS/nixpkgs/pull/217243#issuecomment-1835219430
- An attempt to describe what the interface between Nix and the language should be when it comes to packages
  - https://github.com/NixOS/nix/issues/6507

### Notify maintainers

@infinisil @davhau may already be somewhat familiar with the concepts.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by roberth** [comment]
First attempt:
- https://github.com/NixOS/nixpkgs/pull/296769

**Comment by purepani** [comment]
It might be a good idea to mention the idea of Pure Object Prototypes since that idea was also brought up at one point to solve some of the same problems. I haven't looked into this or POPs enough to know how similar they are so I can't say for certain, but this does seem at least *somewhat* similar from what I know. @roberth sorry for the ping, but is that concept something you know anything about?

**Comment by roberth** [comment]
@purepani No worries! Thanks for highlighting POP. It seems to be a good object system, but the added complexity seems to be a problem, both in terms of contributors having to learn yet another composition mechanism (even if it may be worth learning), and for the possible performance issue raised [here](https://github.com/NixOS/rfcs/pull/91/files#diff-31f36493ab62ba58abd2b11765ebec661b3413e60f0c4872055fe1d3f9243458R362), especially as I do expect it to be amplified by [this](https://github.com/NixOS/rfcs/pull/91/files#diff-31f36493ab62ba58abd2b11765ebec661b3413e60f0c4872055fe1d3f9243458R379).

**Comment by dustyhorizon** [comment]
Really keen to have this implemented, I am maintaining my own internal overlays and set of packages which bears several similarities to `nixpkgs`. 

For python (`python3Packages`) it seems that there are some mechanisms to properly derive to a fixed point; an (inaccurate) example would be me "overriding" a python dependency via `python3Packages.overrideScope ...` and that this overridden dependency seem to be properly resolved in an arbitarily nested python dependency's propagatedBuildInput.

However, I cannot say the same for ocaml (`ocamlPackages`), doing the same `ocamlPackages.overrideScope ...` trick does not properly propagate the overridden package, resulting in a bunch of lines of `Conflicting ocaml packages detected`.

I am looking forward having a standardized overriding mechanism which I am confident in knowing that any overrides will get properly propagated into dependencies (of dependencies of dependencies ...).

**Comment by SomeoneSerge** [comment]
> ```nix
> { deps, ... }:
> {
>   deps = { pkgs, ... }: {
>     /* ... */
>     boost = pkgs.boost_180;
>   };
> ```


Ah, nice! One reading would be that do introduce structured DI hints, but rather than wrapping the lamda into a hints-augmented structure (clumsily described in https://github.com/NixOS/nixpkgs/issues/404946#issuecomment-3314341207), you embed the hints in the lambda using a fixpoint, and you use these "hints" to implement "aliases".

It seems that you choose the "absolute (attribute) paths" approach for the description of how to satisfy the dependencies: rather than asking for a scope-relative "boost_180", you count your implementation path from "pkgs". This probably has implications for relocatability of "recipes", but I am most curious as to how would you approach subscopes such as `python3XXPackages`, where you need to be able to request both `pkgs.{protobuf,cmake,meson}` and `pythonPackages.{protobuf,cmake,meson}`. Would you present them with different global fixpoints?..

**Comment by roberth** [comment]
> One reading would be that this does introduce structured DI hints [...] "aliases".

These are good intuitions. It does a bit more than aliasing, but that's ok.
I think of the expressions like `pkgs.boost_180` in the example as *defaults*.

> you choose the "absolute (attribute) paths" approach

Absolute in a universe where `pkgs` is the origin, but many such universes exist and can be passed into this package function. For a Nixpkgs package, the expectation is of course that it's *a* `pkgs` from the monorepo, of which many exist.

> rather than asking for a scope-relative "boost_180"

It is possible to request `boost_180` as a `deps` dependency and then use it as the default for `boost`.
```nix
{
  deps = { pkgs, boost_180, ... }: {
    inherit (pkgs) boost_180;
    boost = boost_180;
  };
# ...
```

> how would you approach subscopes such as `python3XXPackages`


I'm not sure that a flat scope is desirable.
If the scope is more of an overlay of sorts, it could be modeled as such.
Otherwise, it seems better to model it as an explicit input:

```nix
{
  deps = { pkgs, pythonPkgs, ... }: {
    boost = pkgs.boost_180;
    pythonPkgs = pkgs.python3Packages;
    inherit (pythonPkgs) numpy pandas;
  };
#...
```

Then a scope can be implemented as
- A separate attrset that may (or may not!) be part of `pkgs`
- A mechanism to automatically set (override) a specific dependency attribute (such as `pythonPkgs` here)

In Nixpkgs we'd only need to provide a default for `pythonPkgs` in packages that aren't defined as part of the python package set.

> Would you present them with different global fixpoints?..

Kind of? Their default is somewhat global if it's provided, and realistically those package sets are part of the "global" package set, the `pkgs` fixpoint.

Whether they are their own fixpoints seems like a separate discussion. Since they're part of the `pkgs` fixpoint, they could rely on that fixpoint to find themselves. Might solve some overlay vs package set interactions perhaps, idk.
I try not to redesign everything at once, so that it's more feasible to migrate, but yes, dependency injection and package set structure do interact.

**Comment by SomeoneSerge** [comment]
>  It does a bit more than aliasing, but that's ok.

Definitely. I've been wondering (and have no answers) whether this could afford a better cross-compilation interface than `make-derivaiton.nix` inputs-sorting:

```nix
{
  deps = { pkgs, buildPackages, ... }: {
    wayland-scanner-native = buildPackages.wayland-scanner;
    wayland-scanner-runtime = pkgs.wayland-scanner;
    # ...
  };
  recipe = deps@{ lib, wayland-scanner-native, wayland-scanner-runtime, ... }: {
    # ...
    postPatch = ''
      substituteInPlace ... ... "${lib.getExe wayland-scanner-native}"
    '';
    inherit deps; # no {nativeB,b}uildInputs, sorting already done, without compromising .override:s
  }
}
```

Side note: A day later I feel quite excited about the `{ deps, ... }: { deps = ... ; }` approach and how it almost keeps the signature of "the callpackage-able" the same as it was, but! NixOS ecosystem suggests the thought, and one cannot just dismiss it, that fixpoints and laziness are a dangerous substance which closes many inspection-visualization-diffing-debugging opportunities, compared to a very dumb but static:

```
{
  deps.wayland-scanner-native = {
    offset = "buildHost";
    path = [ "wayland-scanner" ];
    constraints.version.lessThan = "...";
  };
}
```

> I'm not sure that a flat scope is desirable.

Sorry, wdym by a "flat scope"?

> Otherwise, it seems better to model it as an explicit input:
> ```nix
>   deps = { pkgs, pythonPkgs, ... }: { ... }
> ```

Oh I really like that you brought this example here. Currently, all of our `python3XXPackages` do come with a `pythonPackages = self` attribute, which is what, I assume, you reference in the snippet. I was wondering if we could make it a standard that each "fixpoint" expose a reference to itself, and that our DI declarations may feature `pythonPackages`-relative paths, treating `pythonPackages` as a disjoint "root" in its own right. That instead of a bi-directional but linear chain, `pkgs.python3Packages.subScope.foobar` and `[subScope.newScope, python3Packages.newScope, pkgs.newScope]`, we'd have a `pythonPackages` "domain", distinct from "native" `pkgs`. Numpy's recipe's "name" would no longer be `pkgs.python3XXPackages.numpy`, but `pythonPackages.numpy`.

I'm obsessing about the python example because I strongly suspect that whatever design you pick now may, if the PR is merged, unblock us or restrict us further in one particular use-case that is desperately broken in `makeScope*`s:

- enter out-of-tree repo;
- enter native "package recipes", including a library and CLI tool;
- enter python "package recipes";
- python packages depend on the library, each other, and `python3Packages`;
- native packages depend on each other, and `pkgs`;
- CLI tool also depend on python packages;
- "package recipes" be ready for copy-pasting to Nixpkgs;
- "python packages" not depend on python minor version;

This can be hacked together by crafting a custom `newScope` crossing to a sibling subtree, but it's basically a joke.

**Comment by roberth** [comment]
> that fixpoints and laziness are a dangerous substance which closes many inspection-visualization-diffing-debugging opportunities, compared to a very dumb but static:

You could feed it empty data and still get a bunch of static info out. Bringing things into scope by means of a function doesn't have to be a problem.
Meanwhile adding another layer of interpretation (of data this time) adds overhead to an already not super fast system.

> wdym by a "flat scope"?

By flat I meant a single attrset where all (or most) attributes are an individual package.

> we'd have a `pythonPackages` "domain"

I intentionally called it `pythonPkgs` to make it distinct, but I'll leave that up to you and other python maintainers. I suppose `pythonPackages` being Python 2.7 basically means it's up for grabs for this kind of purpose?

This may already be possible today by putting things like `pythonPkgs`/`pythonPackages` in the scope, although I suppose it might be "too breaking", and a clean slate-ish migration to a new style of package definition can be helpful in that regard.

> if the PR is merged, unblock us

Please play around with it. If you can confirm that it helps, that would mean a lot to validate it.
I have approached it as a design cleanup, based on real issues that I know it solves, but I don't know the extent.
Also it can not be a one-person change project anyway. That'd be bad for various reasons.
